docs/create-packages/native-files-in-net-packages.md

```diff
@@ -0,0 +1,88 @@
+
+### Compile assets
+
+Compile assets are passed to the compiler, and therefore must only contain .NET assemblies.
+NuGet will select compile assets from `ref/{tfm}/`, and if that directory does not exist, it will fall back to `lib/{tfm}`, where `{tfm}` is the canonical target framework of the project referencing the package.
+
+The .NET compiler will generate warnings when an assembly targets a different architecture than what the current compilation is targeting, so your package consumers will have the best experience if your `ref/` or `lib/` assemblies target `AnyCPU`.
+
+If your assembly contains a lot of managed code, not just P/Invoke method definitions, you can save space by making the `ref/` assembly a [reference assembly](/dotnet/standard/assembly/reference-assemblies#generating-reference-assemblies), but be sure to include it under `ref/` in your package, not `lib/`.
+
+Therefore, there is no capability to provide different compile-time assemblies for different RIDs, you will need to provide the same API surface for all RIDs, and do runtime checks (for example, throw )
+
+### Runtime assets
+
+Runtime assets are copied to the output directory on build and publish, and the .NET SDK generates the `deps.json` with all the RID-specific asset information.
+NuGet will select runtime assets from `runtimes/{rid}/native/` and `runtimes/{rid}/lib/{tfm}/` (see below to understand `sub-path`), and if that doesn't exist, then it will fall back to `lib/{tfm}/`.
+
+When there are both `runtimes/{rid}/native/` and `runtimes/{rid}/lib/{tfm}/` files, NuGet will always select the `runtimes/{rid}/native/` assets, and then the assets from only one of the `runtimes/{rid}/lib/{tfm}/` directories, just like it does for compile assets in `ref/{tfms}/` or `lib/{tfms}/`.
+
+Similarly, NuGet selects the "best" RID, when multiple exist, or multiple match.
+Therefore, if a package contains both `runtimes/any/native/some.dll` and `runtimes/linux-x64/native/different.so`, when the project consuming the package targets `linux-x64`, NuGet will select `different.so`, and for all other RIDs that are not compatible with `linux-x64`, then NuGet will select `some.dll`.
+You cannot use the `any` RID to put common files that will be selected by all RIDs.
+
+## Putting it together
+
+How to build any binaries native libraries is out of scope of this document.
+It is assumed that all the native binaries and the .NET assemblies have been completed and compied to a machine for packing.
+
+Consider a package `Contoso.Native` that provides .NET APIs to `contoso.dll` on Windows, and `contoso.so` on other platforms.
+The package contents should be (excluding files required by the [Open Packaging Conventions](https://en.wikipedia.org/wiki/Open_Packaging_Conventions)):
+
+```text
+Contoso.Native.1.0.0.nuspec
+ref/net8.0/Contoso.Native.dll
+runtimes/android/lib/net8.0/Contoso.Native.dll
+runtimes/android/native/contoso.so
```

Comment:
```suggestion
runtimes/android/native/libcontoso.so
```

This probably goes for all the `.so` files below, but "linux-like" native libraries will have a `lib` prefix and `.so` file extension.

Reply:
@jonathanpeppers in that case, does the P/Invoke declaration use `[DllImport("contoso")]` or `[DllImport("libcontoso")]`?

Reply:
The underlying linker on Linux systems should search for libraries without the prefix. Thus `contoso` would be the right declaration. IIRC

Reply:
Mono has always let you do any one of:
```csharp
[DllImport("contoso")]
[DllImport("libcontoso")]
[DllImport("libcontoso.so")]
```
We have the same behavior in .NET 6+ MAUI apps today, but I don't offhand know how this works with CoreCLR on Linux.